### ISSUE TITLE
The connect function of TCP doesn't work

### ISSUE THREAD

**Issue by yangsiran** [GaloisInc/HaNS#6]
Running a simple demo like this:
```
import Hans.Address.IP4
import Hans.Device.Xen
import Hans.DhcpClient
import Hans.Layer.Dns(DnsException)
import qualified Hans.NetworkStack as NS

import Network.Socket
import Network.Socket.Internal(setNetworkHansStack)
import Control.Monad
import Control.Concurrent
import System.IO

import Hypervisor.Debug
import Hypervisor.Console
import Hypervisor.XenStore
import XenDevice.NIC

main = do
  con <- initXenConsole

  xs <- initXenStore
  nics <- listNICs xs
  let macstr = head nics
      mac = read macstr

  ns <- NS.newNetworkStack
  nic <- openNIC xs macstr
  NS.addDevice ns mac (xenSend nic) (xenReceiveLoop nic)
  NS.deviceUp ns mac

  ipMV <- newEmptyMVar
  dhcpDiscover ns mac (putMVar ipMV)
  ipaddr <- takeMVar ipMV

  setNetworkHansStack ns

  addr:_ <- getAddrInfo (Just defaultHints) (Just $ show ipaddr) (Just "2222")
  ssock <- socket (addrFamily addr) Stream defaultProtocol
  bind ssock (addrAddress addr)
  listen ssock 5
  writeDebugConsole $ "listen on " ++ show (addrAddress addr) ++ "\n"

  forkIO $ do
    (connsock, caddr) <- accept ssock
    writeDebugConsole $ "connection from " ++ show caddr ++ "\n"
    h <- socketToHandle connsock WriteMode
    hPutStrLn h "hello, world"
    hClose h
    close ssock

  threadDelay 500000
  -- NS.startTcpLayer ns
  csock <- NS.connect ns ipaddr (fromIntegral 2222) Nothing
  writeDebugConsole "csock created\n"

{-
  csock <- socket (addrFamily addr) Stream defaultProtocol
  connect csock (addrAddress addr)
  h <- socketToHandle csock ReadMode
  msg <- hGetLine h
  hClose h
  writeDebugConsole msg
-}
```
The server part, including `bind`, `listen`, `accept` from `network-hans`, and the WebServer example all work, tested using telnet on my build of HaLVM. But the client part, in particular the `connect` function seems to block and can not return. Is my way of calling the socket api wrong ? Or I may just miss some setup step ?

**Comment by elliottt** [comment]
You're right, there's a bug in connect.  The notification that unblocks the calling thread isn't being run, and the code that finalizes the connection is wrong.  I'll try to have a fix up in the next few days :)

**Comment by yangsiran** [comment]
Aha, I have also track down a little. `tcpNotify` which unblock the `MVar` is never called. But things bellow  this layer invoke the implementation of TCP. So I think I should ask you guys first. :)

**Comment by elliottt** [comment]
I believe that this is fixed in the upcoming hans-3.0 release -- connect has been reimplemented, and is simpler and a bit more reliable.
